### PR TITLE
Speed up live reload process

### PR DESCRIPTION
**Pull request by nicholaschuayunzhi**
Follows https://github.com/MarkBind/markbind/pull/128, https://github.com/MarkBind/markbind/pull/144
Related https://github.com/MarkBind/markbind/issues/88

### CODE REVIEW

File: lib/Page.js
Comment (acjh) [inline lib/Page.js]: Why create a `new MarkBind` instance?

Comment (nicholaschuayunzhi) [reply lib/Page.js]: When doing a `serve` command. The page objects created on first build are stored in an array in the site. When changes are made to a file, the same page object is used to regenerate the files.

To be safe, i create a new markbind instance to retain the same behaviour as before. (Where a new page object will be generated together with a new markbind instance).

**Comment by acjh** [comment]
Not working on https://github.com/nus-cs2103/website-base nor a new `markbind init` repo:

screenshot 2018-01-27 03 11 13

**Comment by nicholaschuayunzhi** [comment]
@acjh Did you add the changes in MarkBind/markbind#128 ?

**Comment by acjh** [comment]
Ah right. That's why it's useful to have a meaningful keyword: "Follows MarkBind/markbind#128"

However, there's a huge diff:

**Comment by nicholaschuayunzhi** [comment]
I suspect this could be related https://github.com/MarkBind/markbind/issues/127

Are you able to provide which files you changed??

Also theres no feedback that rebuild has actually finished. I check to see if the .temp file has been cleared.

**Comment by acjh** [comment]
I didn't change any files.

I did `markbind build` on the website without your commit, staged the changes, applied your commit, and then did `markbind build`.

**Comment by nicholaschuayunzhi** [comment]
Im sorry for the late follow up. I have done the same but I have only one diff that comes and goes randomly.

image

I'm not too sure why the :what part seems to render randomly.

Steps i did:
1. `checkout` to master branch for Markbind and Markbind-cli repo
2. make sure to point markbind in markbind-cli to local markbind repo
3. `markbind build` on website-base
4. `commit` the `_site` file
5. `checkout` to dev branch for Markbind and Markbind-cli repo
6. `markbind build` on website-base
7. `git diff`

**Comment by acjh** [comment]
Okay, I couldn't reproduce the diff.

Answering my earlier comment, it makes sense to create a `new MarkBind` instance to invalidate previous `includedFiles`, e.g. when removing an `include`. But this could be done once in the `generate` method.

However, I don't see where `page.includedFiles` is reset. It seems it only adds to it ([L63](https://github.com/nicholaschuayunzhi/markbind-cli/blob/987d7eb653408fed471759f3e53a5aa34b2db0f2/lib/Page.js#L65), [L73](https://github.com/nicholaschuayunzhi/markbind-cli/blob/987d7eb653408fed471759f3e53a5aa34b2db0f2/lib/Page.js#L73)).

**Comment by nicholaschuayunzhi** [comment]
I agree with what you have said. I have not added this feature on file remove yet as I was working on fixing the reload bug which may resolve this issue. Perhaps i'll have to reset the includedFiles to on generate.

**Comment by nicholaschuayunzhi** [comment]
I have rebased this ontop of https://github.com/MarkBind/markbind-cli/pull/21
The site rebuilds correctly for my 2 tests and the steps are:
1. save pregenerated _site
2. make a change to 1 file
3. `$ markbind serve`
4. reverse change in 2 and wait for live reload
5. compare new _site with old site in 1

With these changes i've managed to at least half the wait time for live reload.

This is for a file used in 12 pages
image

This is for a file used in 3 pages
image

Will add functionality for remove and change.
Additionally, the logging info will be in a separate pr.

PS: I don't know why but WebStorm saves some meta files which will trigger the onchange for file watcher.. Advisable to work with a different editor like sublime/atom when testing

**Comment by acjh** [comment]
> 2. make a change to 1 file

When this file is `book/architecture/architecturalStyles/clientServer/what/full.md`, diff in step 5.

**Comment by acjh** [comment]
> diff in step 5.

**Comment by nicholaschuayunzhi** [comment]
Could this be related to the boilerplates file? I can't seem to get that diff.

Update: I'm able to produce a diff when i change the hash in `book/architecture/architecturalStyles/clientServer/what/full.md`

image

**Comment by acjh** [comment]
> Could this be related to the boilerplates file? I can't seem to get that diff.

Ah, I was wondering why I didn't post that file. Here are the relevant screenshots for this issue:

> 2. make a change to 1 file



> diff in step 5.

**Comment by nicholaschuayunzhi** [comment]
Seems like file that is not rebuilding is the same one. Thanks for narrowing down the problem. 👍

**Comment by nicholaschuayunzhi** [comment]
I have addressed the issue in 
https://github.com/MarkBind/markbind/pull/128/commits/33c828bcae5cfebc21969b7e88eaee90e320b7d1

I wasn't adding included file paths for dynamic panels, morphs and panels. Thus the file was not reloaded.

**Comment by acjh** [comment]
Verified.

File: lib/Page.js
Comment (acjh) [inline lib/Page.js]: Looks like this doesn't have to be here. How about a separate `.then` just before `.then(resolve)`?

File: lib/Page.js
Comment (acjh) [inline lib/Page.js]: - This needs a jsdoc.
- Let's name the param as `dependencies`.

File: lib/Page.js
Comment (nicholaschuayunzhi) [reply lib/Page.js]: Should I place it before `resolveDependency` is called. I would prefer to "collect the files" before the recursive method is called.

Comment (acjh) [reply lib/Page.js]: That's independent right?

Collecting just before `.then(resolve)` implies that it doesn't affect the rest of the function. Alternatively, collecting just after `.includeFile` implies that it's unaffected by the rest of the function. These are fine.

On the other hand, collecting just before `resolveDependency` conveys no meaning to other maintainers.

**Comment by nicholaschuayunzhi** [comment]
I've made the requested changes.
EDIT: Still working on some changes.

File: lib/Site.js
Comment (acjh) [inline lib/Site.js]: - Repeated `rebuild`.
- Can you add jsdoc to `buildSourceFiles` and `generatePages` as well, since these are related?

File: lib/Site.js
Comment (acjh) [inline lib/Site.js]: - Move `generateRelevantPages` after `generatePages`.
- Should it be called `regenerateAffectedPages` since:
  - it only `re`generates pages that were previously generated (and consistent with `rebuildSourceFiles`), and
  - it's not obvious what's considered `Relevant`? (`Affected` is marginally better.)

File: lib/Site.js
Comment (nicholaschuayunzhi) [reply lib/Site.js]: Will do!

File: lib/Site.js
Comment (nicholaschuayunzhi) [inline lib/Site.js]: I've changed this to make it consistent with https://github.com/MarkBind/markbind-cli/pull/17/commits/831637cbc93499bb54dbf333e84df866fb479661#diff-e5f3ecb7effaa21db6153e4647ea22bcR197

Rationale is that the method has no knowledge of what is being rebuilt besides the pages. 

Also should we start documenting promises.

File: lib/Site.js
Comment (acjh) [inline lib/Site.js]: "as the source file or as `a` static/dynamic included file"

File: lib/Site.js
Comment (acjh) [inline lib/Site.js]: Remove period.

**Comment by acjh** [comment]
> Also should we start documenting promises.

I'm not sure what you mean by that.

**Comment by nicholaschuayunzhi** [comment]
I meant like methods that return promises should be documented so. At the same time, any things like handling errors should be documented. 
Eg: 
```js
/**
 * @returns {Promise} promise that files are rebuilt, deletes temp and output files on reject
*/ 
```
https://github.com/jsdoc3/jsdoc/issues/1197#issuecomment-312948746
Not too sure how documentation of anonymous promises should be.

**Comment by acjh** [comment]
@Gisonrg Your input please :)

**Comment by nicholaschuayunzhi** [comment]
Change was made in c87d7ba

image

The number logged indicate the number of times file cache was accessed in each markbinder instance. 
Top shows before c87d7ba.
Bottom shows after c87d7ba.

By maintaining the same markbind instance when resolving dependencies, we do not have to access so files from disk.

File: index.js
Comment (Gisonrg) [inline index.js]: Seems like we use these two conditions multiple times, consider extra it to a function in `fsUtil`: `isSourceFile()`

**Review by Gisonrg — CHANGES_REQUESTED** [review]
Code looks fine to me, good job 👍 
Before we merge this, notice that `markbinder.getStaticIncludeSrc()` is not available in the currently used `MarkBind` version. We need to push a minor release to npm before we can use it. `package.json` should also be updated to use the newest version.

We should seriously consider putting two repos together instead of separating them. I will file an issue for this.

File: lib/Site.js
Comment (acjh) [inline lib/Site.js:326]: ~`pages`~`files`

File: lib/Site.js
Comment (acjh) [inline lib/Site.js]: ~`Rebuilt`~`rebuilt`

**Comment by acjh** [comment]
Can you add "Resolves MarkBind/markbind#88" in the PR description?

**Comment by nicholaschuayunzhi** [comment]
Updated!

File: lib/Site.js
Comment (nicholaschuayunzhi) [reply lib/Site.js:326]: I think pages would be more accurate as a page can contain more than one file?

**Comment by nicholaschuayunzhi** [comment]
I have changed `Files Rebuilt` to `Pages rebuilt` to be consistent with `Rebuilding ${processingFiles.length} pages`

File: lib/Site.js
Comment (acjh) [reply lib/Site.js:326]: Hmm okay.

File: lib/util/fsUtil.js
Comment (Gisonrg) [inline lib/util/fsUtil.js]: We don't usually use `module.exports` when exporting functions.
I think if you change to use function object syntax then you can use `this` to reference the functions defined in the object.
```
isSourceFile: function(filePath) {
  return this.isMarkdown(filePath) || this.isHtml(filePath);
}
```
Or use ES6 getter syntax
```
get isSourceFile() {
  return filePath => this.isMarkdown(filePath) || this.isHtml(filePath);
}
```

**Review by Gisonrg — APPROVED** [review]

**Comment by nicholaschuayunzhi** [comment]
I have made the changes requested.

**Comment by nicholaschuayunzhi** [comment]
I have made `collectIncludedFiles` to be placed before the recursive call.
Previously, what was happening was that included files were being re-added to the set as the `dynamicIncludeSrc` and `staticIncludeSrc` was not reset for each recursive call, resulting in slower build.

Hence, I reset at the start of each resolve dependency and `collect` them before the recursive call.

**Comment by acjh** [comment]
Why is resetting necessary and why would it affect the build? Again, it doesn't have to be in the same `.then`.

**Comment by nicholaschuayunzhi** [comment]
If i don't reset the the `dynamicIncludeSrc`, it will keep attempting to rebuild pages that was already built. 

and be caught by:
```js
  if (builtFiles[resultPath]) {
    return resolve();
  }
  // eslint-disable-next-line no-param-reassign
  builtFiles[resultPath] = true;
```

So it works but there's unnecessary execution of code.

EDIT: sorry if i wasn't clear but prior to this, I had changed each execution of generate to use 1 instance of markbind, thus dynamicIncludeSrc and will accumulate all dynamicIncludes and the above scenario will unfold.

**Comment by nicholaschuayunzhi** [comment]
> Again, it doesn't have to be in the same .then

My idea is that since its not asynchronous, we can just have it in that block. However, I can put it to a .then block above if you prefer.

**Comment by acjh** [comment]
> it will keep attempting to rebuild pages that was already built.
>
> I had changed each execution of generate to use 1 instance of markbind, thus dynamicIncludeSrc and will accumulate all dynamicIncludes and the above scenario will unfold.

I still don't see how `builtFiles` affects this.

In fact, with a single instance of markbind, we don't need to collect in `resolveDependency` right?

**Comment by nicholaschuayunzhi** [comment]
> I still don't see how builtFiles affects this.

everytime `this.markbinder.includeFile` is called, it will push dynamic include dependencies into `dynamicIncludeSrc` array. Since we are using the same instance of `markbind` the array persists throughout the recursion. Thus it will recursively call `resolveDependencies` on files that were already included.

Built files prevents the recursion from going indefinitely.

> In fact, with a single instance of markbind, we don't need to collect in resolveDependency right?

If `dynamicIncludeSrc` is reset, then we need to.

**Comment by acjh** [comment]
Ah. That really sheds light on why `resolveDependency` uses a local instance of markbind.

The local instance is cleaner than the resetting workaround.
Let's add your explanation as a comment for clarity though.

**Comment by nicholaschuayunzhi** [comment]
> The local instance is cleaner than the resetting workaround.

Yup is definitely cleaner! But then we have to pay with some IO operations in `markbind`'s `parser` as the filecache is "cleared" when we instantiate a new `markbind` instance.

Would you prefer a local instance, and then a new PR to fix that issue? (I have been working on a static file cache which can be referenced by any instance of markbind)

**Comment by nicholaschuayunzhi** [comment]
This is how it would look like if a local instance were use. I will edit the PR based on the decision made.

File: lib/Page.js
Comment (acjh) [inline lib/Page.js]: `markbinder` can be instantiated after the potential `return` (and `builtFiles[resultPath] = true;`).

As mentioned, let's add your explanation as a comment here.

**Review by acjh — COMMENTED** [review]
Let's go with a local instance. Keeping a reference to `this.markbind` seems unnecessary.

**Comment by acjh** [comment]
Note that this will have to be rebased after #25 is merged, for a clean diff.

**Comment by nicholaschuayunzhi** [comment]
I forgot to note, this is the relevant PR on `markbind` https://github.com/MarkBind/markbind/pull/144 that lets me add boilerplate files to the included files.

**Comment by acjh** [comment]
Noted. Please add "Follows ..." to the PR description.

We need `markbind@1.2.1` before this PR is merged.

**Comment by acjh** [comment]
I've merged #25 and updated to `markbind@1.2.1`.

@nicholaschuayunzhi Can you rebase and test without using a local version of markbind?

**Comment by nicholaschuayunzhi** [comment]
I have tested with a local version of markbind (up to date with upstream master) and the installed node modules version. Im just testing on master branch only.

Want to clarify if markbind@1.2.1 will introduce some formatting changes to the html:
eg:
image

NB: As far as I've checked they are only changes in formatting.

**Comment by acjh** [comment]
That seems to be a result of MarkBind/markbind#131, which was released in `markbind@1.1.7`.

**Comment by nicholaschuayunzhi** [comment]
> That seems to be a result of MarkBind/markbind#131, which was released in markbind@1.1.7.

@acjh That's weird, I should have updated my test site for that. But in any case I have checked that live reload does work for boilerplate files (for my tests) and have rebased accordingly. (With markbind@1.2.1)

**Comment by acjh** [comment]
@nicholaschuayunzhi Thanks for your contribution!

**Pull request merged**
